src/Compilers/CSharp/Portable/Utilities/DeclarationDiagnosticBag.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Takes declaration diagnostics and either stores them directly or acts as a proxy to the compilation's
+    /// <see cref="CSharpCompilation.DeclarationDiagnostics"/>, depending on construction.
+    /// </summary>
+    internal struct DeclarationDiagnosticStore
+    {
+        private readonly DiagnosticBag _backingBagOpt;
+        private readonly CSharpCompilation _compilationOpt;
+
+        public DeclarationDiagnosticStore(DiagnosticBag backingBag)
+        {
+            _backingBagOpt = backingBag ?? throw new ArgumentNullException(nameof(backingBag));
+            _compilationOpt = null;
+        }
+
+        public DeclarationDiagnosticStore(CSharpCompilation compilation)
+        {
+            _compilationOpt = compilation ?? throw new ArgumentNullException(nameof(compilation));
+            _backingBagOpt = null;
+        }
+
+        public void Add(Diagnostic diag) => Bag.Add(diag);
+
+        public void Add(ErrorCode err, Location location) => Bag.Add(err, location);
+
+        public void AddRange(DiagnosticBag diagnostics) => Bag.AddRange(diagnostics);
+
+        public DiagnosticBag Bag
+        {
+            get
+            {
+                var bag = _compilationOpt?.DeclarationDiagnostics ?? _backingBagOpt;
```

Comment:
>_compilationOpt?.DeclarationDiagnostics ?? _backingBagOpt [](start = 26, length = 57)

I don't understand why this entire class can't be replaced by the caller storing the appropriate diagnostic bag.